### ISSUE TITLE
How to compute gradient of loss with respect to input image?

### ISSUE THREAD

**Issue by MatthewInkawhich** [facebookarchive/caffe2#1840]
In Caffe (the original), I was able to obtain the gradient of the loss w.r.t. an input image using the net.backward() function.

		
```
### Load image into net's data layer
img = caffe.io.load_image(jpg_file)
transformed_image = transformer.preprocess('data', img)
net.blobs['data'].data[...] = transformed_image    # copy image data into memory for net

...

### Form the final label vector (one hot vector)
finalLabelVector = np.zeros((1,num_classes))
finalLabelVector[0,ground_truth_label] = 1

### Do a backward pass using finalLabelVector
# must have 'force_backward: true' set in deploy prototxt
backwardpassData = net.backward(**{net.outputs[0]: finalLabelVector})

# Get the derivative with respect to the input (data layer)
delta = backwardpassData['data']
```

After executing this portion, delta contained the gradient of the loss with respect to the input image (and its label).

**How can I achieve this in Caffe2?**

The workspace.RunNet() handles both forward and backward passes in the tutorials that I have read.

Thanks,
Matt

**Comment by harouwu** [comment]
Hi Matt,

Thanks for asking! In caffe2, every Blob is referred by its name. Gradient is also a blob. If you want to access a gradient after some iteration. You can use:


```
g = workspace.FetchBlob(gradient_name)
```


In order to find the blob name, If you are using a `ModelHelper` object to build up your net(model), you can use:


```
model = model_helper.ModelHelper(name="mymodel")
...
for param, grad in model.param_to_grad.items():
     print(param + ', ' + grad)
```


That would help you print out the name of param and gradient blobs

**Comment by MatthewInkawhich** [comment]
@harouwu Thanks for the response!

However, when I run the code that you provide to print the param and grad blobs, I get nothing. In other words, my `model.param_to_grad.items()` is empty.

I am loading the model from a pre-trained model (.pb files), running inference on ONE image, and trying to obtain the gradient of the loss w.r.t. that image. See code below:

```
...
...

arg_scope = {"order": "NCHW"}
test_model = model_helper.ModelHelper(name="test_model", arg_scope=arg_scope, init_params=False)


##################################################################################
# Add the input ('data') stuff to the model (just as in training)
data_uint8, label = test_model.TensorProtosDBInput([], ["data_uint8", "label"], batch_size=1, db=TEST_LMDB, db_type='lmdb')
data = test_model.Cast(data_uint8, "data", to=core.DataType.FLOAT)
data = test_model.Scale(data,data,scale=float(1./256))
# NO STOP GRADIENT AS I WANT GRADIENT OF LOSS WRT INPUT IMAGE (DATA)


##################################################################################
# Add the pretrained model stuff (info from the pbs) to the model helper object
# Populate the model obj with the init net stuff, which provides the parameters for the model
init_net_proto = caffe2_pb2.NetDef()
with open(INIT_NET, "rb") as f:
    init_net_proto.ParseFromString(f.read())
tmp_param_net = core.Net(init_net_proto)
test_model.param_init_net = test_model.param_init_net.AppendNet(tmp_param_net)

# Populate the model obj with the predict net stuff, which defines the structure of the model
predict_net_proto = caffe2_pb2.NetDef()
with open(PREDICT_NET, "rb") as f:
    predict_net_proto.ParseFromString(f.read())
tmp_predict_net = core.Net(predict_net_proto)
test_model.net = test_model.net.AppendNet(tmp_predict_net)


##################################################################################
# Run the test
workspace.RunNetOnce(test_model.param_init_net)
workspace.CreateNet(test_model.net, overwrite=True)
workspace.RunNet(test_model.net.Proto().name)
data = workspace.FetchBlob('data')
softmax = workspace.FetchBlob('softmax')
label = workspace.FetchBlob('label')

print('\ndata:', data[0][0].shape)
print('\nsoftmax: ', softmax[0])
print('\nlabel: ', label[0])

print('params and grads:')
for param, grad in test_model.param_to_grad.items():
     print(param + ', ' + grad)
```

The output to this code is (MNIST net; input image is a 2):
```
WARNING:root:This caffe2 python run does not have GPU support. Will run in CPU only mode.
WARNING:root:Debug message: No module named caffe2_pybind11_state_gpu

data: (28, 28)

softmax:  [  9.91556393e-10   6.10783218e-07   9.99999404e-01   8.50986748e-09
   2.01520287e-10   6.78945994e-13   8.72986197e-11   1.83326532e-09
   7.94001576e-09   6.91495680e-11]

label:  2
params and grads:
```

What am I missing?

**Comment by MatthewInkawhich** [comment]
Ok, I think I figured out how to get the gradient blob I was looking for.

My problem in the last comment was due to the fact that simply bringing up the network from the .pb files is ok if classifying an image using `workspace.Predictor(init_net, predict_net)`.

However, to gain access to gradients, you must "Externally initialize" the params (i.e. load them in the workspace for operation). Also, to get the gradient, you must add gradient operators. See below code for my solution (added this right after loading nets from .pb files and appending to my created net with just a data layer:

```
# Externally initialize params so we can extract gradients
for i,op in enumerate(init_net_proto.op):
	param_name = op.output[0]
	if param_name != 'data':
		assert(op.arg[0].name == "shape")
		tags = (ParameterTags.WEIGHT if param_name.endswith("_w") else ParameterTags.BIAS)
		test_model.create_param(param_name=op.output[0], shape=op.arg[0].ints, initializer=initializers.ExternalInitializer(), tags=tags)


##################################################################################
# Add the "training operators" to the model so we can access model.param_to_grad
xent = test_model.LabelCrossEntropy(['softmax', 'label'], 'xent')
loss = test_model.AveragedLoss(xent, 'loss')
test_model.AddGradientOperators([loss])

# NOW THIS WORKS
print("params and grads:")
for param, grad in test_model.param_to_grad.items():
print(str(param) + ', ' + str(grad))
```

To obtain the data gradient blob, I did not see the gradient blob that I was looking for (I just saw the gradients for the other layers). However, after inspecting the model's Protobuf format (`test_model.net.Proto()`), there was a `data_grad` blob (my data layer was called "data") that can be fetched using `workspace.FetchBlob("data_grad")`.

**Comment by MatthewInkawhich** [comment]
@harouwu 

I have one last question...
Even when I call `test_model.StopGradient(data, data)`, the `data_grad` blob is still created.

Why is this? I thought the StopGradient function would stop the gradient in the backward pass.

**Comment by CriCL** [comment]
Hi @MatthewInkawhich , were you able to input data into an specified layer?
I'm trying to run alexnet with an input image and extract features from fc7 relu.

I'm to close!

**Comment by MatthewInkawhich** [comment]
@CriCL I am not sure if I understand your question. I was not trying to input or inject data into any specified layer, rather, I was trying to obtain a specific gradient blob.

That said, once you run a forward pass on alexnet, it would seem that you would be able to extract the features that you want using `workspace.FetchBlob(<relublob>)`
